### ISSUE TITLE
Slow Category Query

### ISSUE THREAD

**Issue by ImTheRealOne** [bagisto/bagisto#4386]
I have around over 3k in products. All items have a color attribute. We have around 20 colors.
If a category has many different available colors the category page takes 30 to 60+ seconds to load.
I tested out the pull request Duplicated Queries Removed #4315 but performance improved slightly but is still not ideal.

### Preconditions
**Please provide as detailed information about your environment as possible.**

    1. v1.2.0

### Steps to reproduce
**It is important to provide a set of clear steps to reproduce this bug.If relevant please include code samples.**

    1. Import 3k+ items all with a common attribute (i.e. color)
    2. Categorize 3k items in their category
    3. Click on a category with 10+ different possible common attribute value (red, blue, green, etc)

### Expected result
Query is expected to take a little longer but currently taking 30 to 60 seconds plus.

![image](https://user-images.githubusercontent.com/9104404/102851744-7a0f9700-43d1-11eb-8c01-e81ca8f0fbdf.png)
![image](https://user-images.githubusercontent.com/9104404/102851762-809e0e80-43d1-11eb-8588-ddeb173527f9.png)
![image](https://user-images.githubusercontent.com/9104404/102851770-83006880-43d1-11eb-9190-d44284ba9ad6.png)

**Comment by devansh-webkul** [comment]
Hi @ImTheRealOne,

Yes, this is true categories are causing problem. One more thing,

- The PR you mentioned, is just for duplicated queries not for the speed up. Therefore, it will give slightly speed.

- Categories are taking time because categories are loaded on the page header which is render blocking the resources.

**Comment by ImTheRealOne** [comment]
> Hi @ImTheRealOne,
> 
> Yes, this is true. Categories are causing the problem. One more thing,
> 
> * The PR you mentioned, is just for duplicated queries not for the speed up. Therefore, it will give slightly speed.
> * Categories are taking time because categories are loaded on the page header which is render blocking the resources.

Any suggestions on how to improve category page load?

**Comment by devansh-webkul** [comment]
Hi @ImTheRealOne,

Well, this is not advisable to load all the categories on page load, if you check some giant eCommerce site. You will come to know that there is an `All` button or something else. If we click on that button then the categories are loaded.

Right now categories are loaded on page load. I suggest you load the categories on some events after page load. Like hover, click, or something.

Try to breakdown your categories even more. For e.g. just load parent categories than some event like hover or click and then load child one and further.

The best sample, I can give you is the `github.com`. Just open the console and check the network they are fully server-driven. If you hover on your repo branches they still made ajax request.

**Comment by ImTheRealOne** [comment]
I do apologize I did not mean the issue is with the category menu but the product category page. 
![image](https://user-images.githubusercontent.com/9104404/102866093-95879b80-43eb-11eb-9814-5eb48871c10d.png)
![image](https://user-images.githubusercontent.com/9104404/102866100-97e9f580-43eb-11eb-98c8-49889aa13c31.png)

**Comment by devansh-webkul** [comment]
Hi, @ImTheRealOne,

So your homepage is loading fine. The issue is particular on this page.

**Comment by ImTheRealOne** [comment]
yeah the issue is loading all the products on resource/themes/velocity/views/products/index.blade.php

![image](https://user-images.githubusercontent.com/9104404/102871352-2dd54e80-43f3-11eb-842f-85ac4ac2355c.png)

**Comment by devansh-webkul** [comment]
@ImTheRealOne,

I have marked this as optimization. I will let you know if this is completed. Once done you can use that PR.

**Comment by ImTheRealOne** [comment]
@devansh-webkul 

I was able to optimize the product attribute filter by replacing the double for loop code in ProductFlatRepository in
 public function getProductsRelatedFilterableAttributes($category) {


       `$allProductAttributeOptionsCode = [];
        foreach ($products as $key => $product) {
            foreach ($filterAttributes as $attribute) {
                if ($attribute->code <> 'price' && isset($product[$attribute->code])) {
                    if (! in_array($product[$attribute->code], $allProductAttributeOptionsCode)) {
                        array_push($allProductAttributeOptionsCode, $product[$attribute->code]);
                    }
                }
            }`
  
with


            $allProductAttributeOptionsCode = [];
        foreach ($filterAttributes as $attribute) {
            if ($attribute->code <> 'price') {
                $allProductAttributeOptionsCode = array_merge($allProductAttributeOptionsCode, $products->unique($attribute->code)->whereNotNull($attribute->code)->pluck($attribute->code)->toArray());
            }
        }


I am having issues trying to find where the product is fetching every attribute?
![image](https://user-images.githubusercontent.com/9104404/102970994-65083600-44ad-11eb-99c2-4b89a20656de.png)

**Comment by devansh-webkul** [comment]
Hi @ImTheRealOne,

It would be nice if you make a PR. We can test it on our end as well.

**Comment by krishanujadiya** [comment]
We were doing some load tests with 5 lac products and 4k categories.
After randomly linking the products to these categories we now have:

- 0.5M records in product_flat
- 1M records in product_categories
- 8M records in product_attribute_values

Now there is a huge delay of 30 secs to execute a query which is even called multiple times. I know there is an issue raised to resolve that and 2 queries have been removed.

But after analyzing the database here our my thoughts, I may be wrong but please correct me:
- If we can move the meta_* columns to another table that is only used in the product detail page then we can save 2M records in product_attribute_values.

- Why are we saving product textual content in product_attribute_values since it is already there in product_flat. I know we have an option to compare attributes but we do need meta_title, meta_keyword, meta_description from attributes.

- Ideally, the only filterable attribute should be saved in product_attribute_values like colour, size, brands, price, etc

**Comment by mottaviani-dev** [comment]
From what I know this strategy of data should work for a small number of products (10K) but here the situation is getting ridiculous at 300 products with 20 attributes.


https://github.com/bagisto/bagisto/blob/master/packages/Webkul/Product/src/Repositories/ProductRepository.php#L156

this query is a disaster and I just replaced it with this

```
  $qb = $query
      ->select('product_flat.*')
      ->join('product_categories', 'product_categories.product_id', '=', 'product_flat.product_id')
      ->join('product_attribute_values', 'product_attribute_values.product_id', '=', 'variants.product_id')
      ->where('product_flat.channel', $channel)
      ->where('product_flat.locale', $locale)
      ->whereNotNull('product_flat.url_key');
```

there is a distinct (that may be a huge hack), a double join between product_flat and himself aliased with 'variants'  ( we are in the function getALL) I didn't check because I trust that the ProductFlatRepository is already returning all the flats I don't need to rejoin them, but in reality, the real question here is why this endpoint is full scanning the database to display 9 products.

I also don't see the usage of indexes on any of these where condition columns but maybe just my editor.

While writing this I imported 

![Screenshot 2021-04-25 at 13 52 57](https://user-images.githubusercontent.com/83165010/115992355-96188480-a5cd-11eb-9d32-99437fb47bc5.png)

resulting in 2.5ish seconds for the busy endpoint

![Screenshot 2021-04-25 at 13 54 03](https://user-images.githubusercontent.com/83165010/115992394-ce1fc780-a5cd-11eb-91d0-0f375fb8c6d7.png)

il keep you posted if I get any benefit from adding indexes

Also if I will realize why there was the double Join with `product_flat`  I may come back to correct myself :D

**Comment by devansh-webkul** [comment]
Hi @mottaviani-dev,

The problem is with the `product_attribute_values` table and from indexing only display record performance can be increased but for the category listing page, the problem will still there. This table should be avoided rather than multiple joining you can use one `product_flat` table.

In current branch we haven't done that but we will look into it later.

**Comment by q8groups** [comment]
this query is disaster, either you will fix it by using elasticsearch or changing this, its taking for me more than 7000 seconds
<img width="1406" alt="Screen Shot 2022-05-17 at 12 51 15 AM" src="https://user-images.githubusercontent.com/3493885/168688542-a3ad3562-ea56-4719-83ce-6379914c83dc.png">

**Comment by jcprangue** [comment]
I am having the same issue with this.. is this issue is fixed? 
thanks

**Comment by jitendra-webkul** [comment]
Fixed in #6422
